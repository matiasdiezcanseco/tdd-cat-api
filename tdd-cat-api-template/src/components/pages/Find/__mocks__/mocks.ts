export const mockCat = {
  name: 'Abyssinian',
  imgUrl: 'https://cdn2.thecatapi.com/images/itfFA4NWS.jpg',
  width: 914,
  heigth: 1280,
}
