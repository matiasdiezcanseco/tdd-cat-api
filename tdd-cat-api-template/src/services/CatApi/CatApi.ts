import axios from 'axios'

export const getCatImageByBreedId = async (id: string) => {}
