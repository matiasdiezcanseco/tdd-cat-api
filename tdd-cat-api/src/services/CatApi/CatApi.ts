import axios from 'axios'

export const getCatImageByBreedId = async (id: string) => {
  try {
    const res = await axios.get('https://api.thecatapi.com/v1/images/search', {
      params: {
        breed_id: id,
        size: 1,
        limit: 1,
      },
    })
    const data = res.data
    return {
      name: data[0].breeds[0].name,
      url: data[0].url,
      height: data[0].height,
      width: data[0].width,
    }
  } catch (e) {
    return {
      name: '',
      url: '',
      height: 0,
      width: 0,
    }
  }
}
