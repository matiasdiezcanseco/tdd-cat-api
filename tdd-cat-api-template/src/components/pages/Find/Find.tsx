import { useEffect, useState } from 'react'
import { getCatImageByBreedId } from '../../../services/CatApi/CatApi'
import Button from '../../atoms/Button/Button'
import Input from '../../atoms/Input/Input'
import CatDisplay from '../../molecules/CatDisplay/CatDisplay'
import CatList from '../../molecules/CatList/CatList'
import Layout from '../../templates/Layout/Layout'

const Find = () => {
  return (
    <Layout>
      <div className="flex justify-between p-2 space-y-2" data-testid="findpage-testid">
        <div className="space-y-2"></div>
        <div></div>
      </div>
    </Layout>
  )
}

export default Find
