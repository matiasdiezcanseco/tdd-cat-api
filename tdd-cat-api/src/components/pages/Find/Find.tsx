import { useState } from 'react'
import { getCatImageByBreedId } from '../../../services/CatApi/CatApi'
import Button from '../../atoms/Button/Button'
import Input from '../../atoms/Input/Input'
import CatDisplay from '../../molecules/CatDisplay/CatDisplay'
import CatList from '../../molecules/CatList/CatList'
import Layout from '../../templates/Layout/Layout'

const Find = () => {
  const [loading, setLoading] = useState(false)
  const [search, setSearch] = useState('')
  const [cat, setCat] = useState<{
    name: any
    url: any
    height: any
    width: any
  } | null>(null)

  const handleApiCall = async () => {
    setLoading(true)
    const _cat = await getCatImageByBreedId(search)
    console.log(_cat)
    setCat(_cat)
    setLoading(false)
  }

  return (
    <Layout>
      <div className="flex justify-between p-2 space-y-2" data-testid="findpage-testid">
        <div className="space-y-2">
          <Input onChange={(e) => setSearch(e)} title="Buscar por id: " />
          <Button text="Buscar" onClick={handleApiCall} loading={loading} />
          {cat?.name && <CatDisplay cat={cat} />}
        </div>
        <div>
          <CatList />
        </div>
      </div>
    </Layout>
  )
}

export default Find
