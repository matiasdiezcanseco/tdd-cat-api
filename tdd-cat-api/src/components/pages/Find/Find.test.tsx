import { render, screen, waitFor } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { BrowserRouter } from 'react-router-dom'
import { mockCat } from './__mocks__/mocks'
import Find from './Find'
import * as catApi from '../../../services/CatApi/CatApi'

// jest.mock('../../../services/CatApi/CatApi', () => {
//   return {
//     getCatImageByBreedId: () => mockCat as any,
//   }
// })

jest.mock('../../../services/CatApi/CatApi')
const mockCatApi = catApi as jest.Mocked<typeof catApi>

describe('test Find', () => {
  beforeEach(() => {
    mockCatApi.getCatImageByBreedId.mockResolvedValue(mockCat as any)
  })

  test('should show cat after api is called ', async () => {
    render(
      <BrowserRouter>
        <Find />
      </BrowserRouter>
    )
    const button = screen.getByRole('button')
    await userEvent.click(button)
    const catName = await screen.findByTestId('catname-testid')
    const catImage = await screen.findByRole('img')
    expect(catName).toBeVisible()
    expect(catImage).toBeVisible()
  })

  test('should not be loading after api call is finished', async () => {
    render(
      <BrowserRouter>
        <Find />
      </BrowserRouter>
    )
    const button = screen.getByRole('button')
    await userEvent.click(button)
    expect(screen.queryByTestId('loading-testid')).toBeNull()
  })

  test('should show typed text', async () => {
    render(
      <BrowserRouter>
        <Find />
      </BrowserRouter>
    )
    const textbox = screen.getByRole('textbox')
    await userEvent.type(textbox, 'test')
    expect(textbox).toHaveValue('test')
  })

  test('should match snapshot', () => {
    const { container } = render(
      <BrowserRouter>
        <Find />
      </BrowserRouter>
    )
    expect(container).toMatchSnapshot()
  })
})
