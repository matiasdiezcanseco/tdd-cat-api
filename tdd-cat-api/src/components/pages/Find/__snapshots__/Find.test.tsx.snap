// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test Find should match snapshot 1`] = `
<div>
  <div
    class="flex flex-col h-screen"
  >
    <div
      class="py-4 bg-slate-500"
    >
      <ul
        class="flex justify-center space-x-40 text-gray-100 text-lg"
      >
        <li
          class="cursor-pointer hover:underline"
        >
          Home
        </li>
        <li
          class="cursor-pointer hover:underline"
        >
          Buscar
        </li>
      </ul>
    </div>
    <div
      class="flex-grow bg-gray-300"
    >
      <div
        class="flex justify-between p-2 space-y-2"
        data-testid="findpage-testid"
      >
        <div
          class="space-y-2"
        >
          <label>
            Buscar por id: 
          </label>
          <input
            class="border-[1px] rounded-md border-gray-600 p-2 bg-transparent"
            type="text"
          />
          <button
            class="min-w-[80px] flex justify-center p-2 text-gray-200 rounded-md bg-slate-600"
          >
            Buscar
          </button>
        </div>
        <div>
          <div
            class="py-2 px-4 border-[1px] space-y-2 border-gray-600 rounded-md"
          >
            <h1
              class="text-2xl text-center"
            >
              Lista de ids
            </h1>
            <ul
              class="space-y-1"
            >
              <li
                class=""
              >
                Abyssinian: abys
              </li>
              <li
                class=""
              >
                American Bobtail: abob
              </li>
              <li
                class=""
              >
                American Curl: acur
              </li>
              <li
                class=""
              >
                Australian Mist: abys
              </li>
              <li
                class=""
              >
                Bengal: beng
              </li>
              <li
                class=""
              >
                Bombay: bomb
              </li>
              <li
                class=""
              >
                Chartreux: char
              </li>
              <li
                class=""
              >
                Cheetoh: chee
              </li>
              <li
                class=""
              >
                Cymric: cymr
              </li>
            </ul>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
